### ISSUE TITLE
deeplearning - 5/w3

### ISSUE THREAD

**Issue by getElementsByName** [getElementsByName/tmp-doc#8]
# 5 Sequence Models - 3 Sequence models & Attention mechanism
- Encoder-decoder model / Sequence to sequence model
- Beam Search
- Bleu Score
- Attention Model
- spectrogram

### Basic Models
- `Encoder-decoder model`
- `Sequence to sequence model` 예
   -  French -> English translation
![image](https://user-images.githubusercontent.com/21363169/89138519-d5ea9800-d576-11ea-8525-ce1e1fd47a34.png)
- image-to-sequence or image captioning model 예
   - Image Captioning
![image](https://user-images.githubusercontent.com/21363169/89138530-dbe07900-d576-11ea-83e1-6f17fad93dae.png)
-  you `don't` want a `randomly` chosen translation, you maybe `want` `the most likely` translation 

### Picking the most likely sentence
- `conditional language model`
![image](https://user-images.githubusercontent.com/21363169/89139075-8c9b4800-d578-11ea-9376-558456b231d2.png)
![image](https://user-images.githubusercontent.com/21363169/89140303-33cdae80-d57c-11ea-8a12-1eeb32d3ef69.png)
![image](https://user-images.githubusercontent.com/21363169/89140309-3c25e980-d57c-11ea-963a-58a53d9338f9.png)

### Beam Search
![image](https://user-images.githubusercontent.com/21363169/89141024-4cd75f00-d57e-11ea-8df5-51621a1a8e8a.png)
![image](https://user-images.githubusercontent.com/21363169/89141171-b6576d80-d57e-11ea-88de-9d3f0ab4c262.png)

### Refinements to Beam Search
- Sometimes it's called `a normalized log` likelihood objective.
- underflow
   -> ~~production~~ -> log (sum)
-  the more terms you have together, the more negative this thing becomes.
   -> normalize this by `the number of words` in your translation
        - And this significantly reduces the penalty for outputting longer translations.
- beam width가 1~10에서 효과가 클 것이다
![image](https://user-images.githubusercontent.com/21363169/89141229-e3a41b80-d57e-11ea-89af-3efdc74468b3.png)
![image](https://user-images.githubusercontent.com/21363169/89141240-f1f23780-d57e-11ea-8e76-137b88906c5e.png)

### Error analysis in beam search
- your model has two main components.
    - neural network model (seq2seq, RNN)
    - Beam search algorithm
![image](https://user-images.githubusercontent.com/21363169/89141623-10a4fe00-d580-11ea-9c5e-e650a2ebd571.png)

### Bleu Score (optional)
- BLEU(Bilingual Evaluation Understudy)
   - the BLEU score is an understudy, could be a substitute for having humans evaluate every output of a machine translation system.

![image](https://user-images.githubusercontent.com/21363169/89142387-37fcca80-d582-11ea-9ec2-e1d7bf748154.png)
![image](https://user-images.githubusercontent.com/21363169/89142396-3fbc6f00-d582-11ea-9d3b-2be588482078.png)
- BP, Stands for brevity penalty
    - machine translation system이 짧은 문장의 결과를 내었을때, 의도치 않게 정확도가 높아지는 문제를 보정
![image](https://user-images.githubusercontent.com/21363169/89142574-c6714c00-d582-11ea-8087-95425ad7a802.png)

#### 기존 구조의 한계
- [X] for very long sentences, maybe longer than 30 or 40 words, the performance comes down
    - encoder의 결과를 모두 기억했다가 decoder로 한번에 전달
- [ ] short sentences are just hard to translate, hard to get all the words
![image](https://user-images.githubusercontent.com/21363169/89142778-59aa8180-d583-11ea-8201-ff65655c4640.png)

#### Attention Model 아이디어
- Encoder의 내용들에 각각 가중치를 적용하여 decoder에 반영
![image](https://user-images.githubusercontent.com/21363169/89142981-0684fe80-d584-11ea-8fc6-596e004b09dc.png)

### Attention Model
- c: context vector
![image](https://user-images.githubusercontent.com/21363169/89143237-d8ec8500-d584-11ea-9b33-518da0a4838a.png)
- e: energies
![image](https://user-images.githubusercontent.com/21363169/89143246-e43fb080-d584-11ea-924c-3da29ab76371.png)

### Speech recognition
- pre-processing 
   - `spectrograms`
       - 주파수와 진폭(세기)정보를 시간의 흐름에 따라 표현
       - Short-time Fourier Transform 의 결과로 표현
   - 달팽이관 구조 https://brunch.co.kr/@cogsciin/45
![image](https://user-images.githubusercontent.com/21363169/89143549-ec4c2000-d585-11ea-86c8-5c2700cbdb1b.png)

![image](https://user-images.githubusercontent.com/21363169/89143613-2a494400-d586-11ea-8716-e1c2d1b09601.png)

- 음성데이터에 대해 생각할 수 있는 한가지 neural network model
![image](https://user-images.githubusercontent.com/21363169/89144010-7fd22080-d587-11ea-83c5-b149e35f0758.png)

- CTC (Connectionist temporal classification)
    - input과 output의 time step 수를 동일하게 하는 방안
![image](https://user-images.githubusercontent.com/21363169/89143845-e6a30a00-d586-11ea-8df2-8ce01854b064.png)
![image](https://user-images.githubusercontent.com/21363169/89143944-48637400-d587-11ea-801d-1c2bea343547.png)

### Trigger Word Detection
- trigger word detection (sometimes also called keyword detection, or wake word detection)
  - Trigger word detection is the technology that allows devices like Amazon Alexa, Google Home, Apple Siri, and Baidu DuerOS to wake up upon hearing a certain word.

- y_hat<t> when that that clip has finished saying "activate".
- Rather than updating the label of a single time step, we will update several(50) steps of the output to have target label 1.
   -  That updating several consecutive time steps can make the training data more balanced.
![image](https://user-images.githubusercontent.com/21363169/89144408-aa70a900-d588-11ea-8171-018872d3b4d0.png)

**Comment by getElementsByName** [comment]
### Do we need to use beam search in training process?
https://stackoverflow.com/questions/44228271/do-we-need-to-use-beam-search-in-training-process


### transformer
- https://www.tensorflow.org/tutorials/text/transformer
- https://ai.googleblog.com/2017/08/transformer-novel-neural-network.html
